modules/charset/charset.go

```diff
@@ -18,11 +19,215 @@ import (
 	"github.com/gogs/chardet"
 	"golang.org/x/net/html/charset"
 	"golang.org/x/text/transform"
+	"golang.org/x/text/unicode/bidi"
 )
 
 // UTF8BOM is the utf-8 byte-order marker
 var UTF8BOM = []byte{'\xef', '\xbb', '\xbf'}
 
+// EscapeStatus represents the findings of the unicode escaper
+type EscapeStatus struct {
+	Escaped      bool
+	HasError     bool
+	HasBadRunes  bool
+	HasControls  bool
+	HasSpaces    bool
+	HasMarks     bool
+	HasBIDI      bool
+	BadBIDI      bool
+	HasRTLScript bool
+	HasLTRScript bool
+}
+
+// Or combines two EscapeStatus structs into one representing the conjunction of the two
+func (status EscapeStatus) Or(other EscapeStatus) EscapeStatus {
+	status.Escaped = status.Escaped || other.Escaped
+	status.HasError = status.HasError || other.HasError
+	status.HasBadRunes = status.HasBadRunes || other.HasBadRunes
+	status.HasControls = status.HasControls || other.HasControls
+	status.HasSpaces = status.HasSpaces || other.HasSpaces
+	status.HasMarks = status.HasMarks || other.HasMarks
+	status.HasBIDI = status.HasBIDI || other.HasBIDI
+	status.BadBIDI = status.BadBIDI || other.BadBIDI
+	status.HasRTLScript = status.HasRTLScript || other.HasRTLScript
+	status.HasLTRScript = status.HasLTRScript || other.HasLTRScript
+	return status
+}
+
+// EscapeControlString escapes the unicode control sequences in a provided string and returns the findings as an EscapeStatus and the escaped string
+func EscapeControlString(text string) (EscapeStatus, string) {
+	sb := &strings.Builder{}
+	escaped, _ := EscapeControlReader(strings.NewReader(text), sb)
+	return escaped, sb.String()
+}
+
+// EscapeControlBytes escapes the unicode control sequences  a provided []byte and returns the findings as an EscapeStatus and the escaped []byte
+func EscapeControlBytes(text []byte) (EscapeStatus, []byte) {
+	buf := &bytes.Buffer{}
+	escaped, _ := EscapeControlReader(bytes.NewReader(text), buf)
+	return escaped, buf.Bytes()
+}
+
+// EscapeControlReader escapes the unicode control sequences  a provided Reader writing the escaped output to the output and returns the findings as an EscapeStatus and an error
+func EscapeControlReader(text io.Reader, output io.Writer) (escaped EscapeStatus, err error) {
+	buf := make([]byte, 4096)
+	readStart := 0
+	var n int
+	var writePos int
+
+	lineHasBIDI := false
+	lineHasRTLScript := false
+	lineHasLTRScript := false
+
+readingloop:
+	for err == nil {
+		n, err = text.Read(buf[readStart:])
```

Comment:
Can we use `ReadBytes` with delimiter here? Then we can process line by line, and we do not need to handle edge cases like `runes can be at most 4 bytes - so...`. Since we finally read all content into a string for FileContent, here we use ReadBytes won't make memory worse.

Reply:
If you really wanted to simplify, using an io.RuneReader would be the most simple thing but then you're going to end up writing to the output very frequently.

This implementation just uses a simple io.Reader and io.Writer it's not dependent on having to have the whole thing in memory or not.